src/main.rs

```diff
@@ -322,3 +337,64 @@ fn apply_suggestion(suggestion: &Suggestion) -> Result<(), ProgramError> {
 
     Ok(())
 }
+
+
+/// We need some special care for use suggestions,
+/// to eliminate duplicates, for example for HashMap,
+/// rustc suggests 4 variants:
+/// use std::collections::HashMap;
+/// use std::collections::hash_map::HashMap;
+/// use std::collections::HashMap;
+/// use std::collections::hash_map::HashMap;
+///
+/// Also it would be good to place them at the top of file
+fn comb_use_suggestions(suggestions: &mut Vec<Suggestion>) {
+    let mut to_remove = HashSet::new();
+    {
+        type UseSuggestionId<'a> = (&'a str, LineRange, &'a str);
+        let mut use_classes = HashMap::<UseSuggestionId, Vec<usize>>::new();
+        for (i, s) in suggestions.iter().enumerate() {
+            if s.is_use_suggestion() {
+                let short_type_name = {
+                    let use_expr = &s.replacement;
+                    assert!(use_expr.starts_with("use "));
+                    assert!(use_expr.ends_with(";\n"));
+                    use_expr.rfind("::").map(|idx|
+                                             &use_expr[idx + 2..use_expr.len() - 2])
+                        .unwrap_or(&use_expr[4..use_expr.len() - 2])
+                };
+                let id = (s.file_name.as_str(), s.line_range, short_type_name);
+                use_classes.entry(id).or_insert(vec![i]).push(i);
+            }
+        }
+
+        for (_, class) in &use_classes {
+            let min_pos: usize = class[class
+                                       .iter()
+                                       .map(|v| suggestions[*v].replacement.len())
+                                       .enumerate()
+                                       .min_by(|l, r| l.1.cmp(&r.1))
+                                       .unwrap().0];
+            for idx in class.iter() {
+                if *idx != min_pos {
+                    to_remove.insert(*idx);
+                }
+            }
+        }
+    }
+
+    let ret: Vec<_> = suggestions.drain(..)
+        .enumerate()
+        .filter(|&(i, _)| !to_remove.contains(&i))
+        .map(|v| {
+            let new_v = Suggestion {
+                line_range:  LineRange {
+                    start: LinePosition { line: 1, column: 1},
```

Comment:
We should not be hacking around with the insertion span. Let's fix https://github.com/rust-lang/rust/issues/42835 instead.